firebase-firestore/src/jsMain/kotlin/dev/gitlive/firebase/firestore/internal/NativeFirebaseFirestoreWrapper.kt

```diff
@@ -20,28 +21,41 @@ import kotlinx.coroutines.GlobalScope
 import kotlinx.coroutines.await
 import kotlinx.coroutines.promise
 
+// There is currently no way to check whether Firestore was already initialized for a given app without actually initializing it
+// Therefore we keep track of this internally
+private val appsWithFirestore = mutableListOf<FirebaseApp>()
+
```

Comment:
what are we doing exactly that is straightforward on android but  not supported natively on ios and js? The other option here would be to cater for the lowest common denominator, in other words don't support on android either,  but I still not exactly clear what the differences are?

Reply:
So there are two things:

First and foremost, the biggest difference between the mobile platforms and JS is that for the mobile platforms, the FirebaseFirestore object internally keeps a state machine that allows it to change settings until it is first used. So, you can do:

```
firestore.settings.host = "1.1.1.1"
firestore.enableNetwork()
firstore.settings.host = "2.2.2.2" // This is not allowed
```

Javascript on the other hand has no such stateMachine with regards to settings. You can either get a FirebaseFirestore instance with whatever settings it was previously initialized, or you can initialize it with settings provided it has not been initialized before (there's no telling whether it is btw). The only one you can do is enable the emulator, which you can do after initialization but only before doing any operations.

This difference is the main reason for this major discrepancy. The approach of this library has always been to get the FirebaseFirestore object and change its settings before operations are executed. If you look at the `setSettings` method on JS pre 1.13.0, you'll realize that it would've crashed if you called the method as it calls the initializationCode after it has already gotten an initialized instance.

I guess it __might__ be possible to essentially remove the state machine logic out of the library, so that you must provide the settings when getting a FirebaseFirestore instance, but this might have unknown side effects and is a breaking change for sure.

The second problem is smaller, and could be solved by omission: Settings on Android are fully readable, on iOS readable except for the cache settings, and on JS (due to initialization mentioned above) not readable at all. As of this PR, we get the settings on Android/iOS (though we miss a few cache settings on iOS) and we track it ourselves for JS.

Reply:
Ok well for the second issue I think removing the getter so you can't read settings on any platform is sensible.

As for the first issue, looking at https://firebase.google.com/docs/firestore/manage-data/enable-offline#configure_offline_persistence, I suggest the following: 
-  make it only possible to set all the settings at once so `firestore.settings.host = "1.1.1.1` shouldn't be possible (which will be the case once the getter is removed)
- on JS, since `Firebase.firestore.settings = firestoreSettings { ... }` needs to call `initializeFirestore` before `getFirestore` is called can the js property be lazy so `getFirestore` is not called doing  `Firebase.firestore.settings = firestoreSettings { ... }`
- if the user attempts to set the settings after doing another operation which means `getFirestore` has already been called then let the js sdk (and others) throw the relevant error.

Reply:
Agreed:

- Removed settings property in favour of `setSettings(builder)` and `applySettings(settings)` methods. This is an api change compared to 1.13.0 but I think we're both fine with that.
- The JS implementation now triggers "setting" the settings if the firestore object has been initialized. This should ensure you get a native crash if you try and set settings after already having initialized the firestore object.

Reply:
Lets keep the syntax matching the [android sdk](https://github.com/firebase/snippets-android/blob/bdab1d019d41c1e034ecc1d63d30c7071af3d79d/firestore/app/src/main/java/com/google/example/firestore/kotlin/DocSnippets.kt#L112-L118) with a [setter only property](https://stackoverflow.com/a/54983086/737857) rather that two new functions

Reply:
Its not an officially supported feature, so it has some downsides as your link shows. Regardless, Ive implemented it as requested